File: 01-scenarios/Web-App-Agent-User-OBO/0.Resources/Starting-point/python/app/main.py
"""FastAPI host for the Microsoft Learn agent."""

from __future__ import annotations

import logging
from contextlib import asynccontextmanager
from pathlib import Path

from fastapi import FastAPI, HTTPException, Request, status
from fastapi.responses import FileResponse
from fastapi.staticfiles import StaticFiles
from pydantic import BaseModel, Field

from app.agent import LearnAgent
from app.auth import AuthRequiredError, AuthService
from app.config import settings

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(name)s: %(message)s")
logger = logging.getLogger("learn-agent")

STATIC_DIR = Path(__file__).parent / "static"

agent = LearnAgent(settings)
auth_service = AuthService(settings) if settings.entra_sign_in_enabled else None


@asynccontextmanager
async def lifespan(_: FastAPI):
    if auth_service:
        logger.info(
            "Entra sign-in is configured; requesting user tokens for %s.",
            settings.agent_blueprint_scope,
        )
    else:
        logger.info(
            "Entra sign-in is not configured; running anonymously. Fill in the AzureAd settings in .env to enable it."
        )
    # Connect to the Microsoft Learn MCP server and discover its tools once at startup,
    # so every chat turn reuses the same tool list.
    await agent.start()
    logger.info("Listening on http://%s:%d", settings.host, settings.port)
    yield


app = FastAPI(title="Microsoft Learn Agent", lifespan=lifespan)
app.mount("/static", StaticFiles(directory=STATIC_DIR), name="static")
if auth_service:
    app.include_router(auth_service.router)


class ChatRequest(BaseModel):
    session_id: str = Field(min_length=1)
    message: str = Field(min_length=1)


class ChatResponse(BaseModel):
    reply: str


@app.get("/")
async def index() -> FileResponse:
    return FileResponse(STATIC_DIR / "index.html")


@app.get("/api/info")
async def info() -> dict[str, object]:
    return {
        "deployment": settings.azure_openai_deployment,
        "learnMcpEndpoint": settings.learn_mcp_endpoint,
        "tools": agent.tool_names,
    }


if not auth_service:

    @app.get("/api/me")
    async def me() -> dict[str, object]:
        return {
            "authenticationConfigured": False,
            "authenticated": False,
            "user": None,
        }


@app.post("/api/chat")
async def chat(request: ChatRequest, http_request: Request) -> ChatResponse:
    if auth_service:
        try:
            auth_service.acquire_user_assertion(http_request)
        except AuthRequiredError as error:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=str(error),
            ) from error

    try:
        reply = await agent.ask(request.session_id, request.message)
    except Exception:
        logger.exception("Chat turn failed")
        return ChatResponse(
            reply="Sorry, something went wrong while researching that. Please try again."
        )

    return ChatResponse(reply=reply)


def run() -> None:
    import uvicorn

    uvicorn.run("app.main:app", host=settings.host, port=settings.port)


if __name__ == "__main__":
    run()

File: 01-scenarios/Web-App-Agent-User-OBO/0.Resources/Starting-point/nodejs/src/main.ts
/** Express host for the Microsoft Learn agent. */

import { fileURLToPath } from "node:url";
import path from "node:path";

import express from "express";

import { LearnAgent } from "./agent.js";
import { AuthRequiredError, AuthService } from "./auth.js";
import { settings } from "./config.js";

const PUBLIC_DIR = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "public");

const agent = new LearnAgent(settings);
const authService = settings.entraSignInEnabled ? new AuthService(settings) : undefined;

const app = express();
app.use(express.json());
app.use("/static", express.static(PUBLIC_DIR));

if (authService) {
    app.use(authService.router);
    console.info(`Entra sign-in is configured; requesting user tokens for ${settings.agentBlueprintScope}.`);
} else {
    console.info(
        "Entra sign-in is not configured; running anonymously. Fill in the AZURE_AD settings in .env to enable it.",
    );

    app.get("/api/me", (_req, res) => {
        res.json({ authenticationConfigured: false, authenticated: false, user: null });
    });
}

app.get("/", (_req, res) => {
    // The file is resolved relative to `root` so that a dot-prefixed folder anywhere in the
    // absolute path - C:\Users\you\.something\ - isn't treated as a hidden file and 404'd.
    res.sendFile("index.html", { root: PUBLIC_DIR });
});

app.get("/api/info", (_req, res) => {
    res.json({
        deployment: settings.azureOpenAiDeployment,
        learnMcpEndpoint: settings.learnMcpEndpoint,
        tools: agent.toolNames,
    });
});

app.post("/api/chat", async (req, res) => {
    const { session_id: sessionId, message } = req.body ?? {};

    if (typeof sessionId !== "string" || !sessionId || typeof message !== "string" || !message) {
        res.status(400).json({ reply: "Both session_id and message are required." });
        return;
    }

    if (authService) {
        try {
            await authService.acquireUserAssertion(req);
        } catch (error) {
            if (error instanceof AuthRequiredError) {
                res.status(401).json({ reply: error.message });
                return;
            }

            console.error("Could not acquire the user assertion", error);
            res.status(401).json({ reply: "Sign in before chatting with the agent." });
            return;
        }
    }

    try {
        const reply = await agent.ask(sessionId, message);
        res.json({ reply });
    } catch (error) {
        console.error("Chat turn failed", error);
        res.json({ reply: "Sorry, something went wrong while researching that. Please try again." });
    }
});

// Connect to the Microsoft Learn MCP server and discover its tools once at startup,
// so every chat turn reuses the same tool list.
await agent.start();

app.listen(settings.port, settings.host, () => {
    console.info(`Listening on http://${settings.host}:${settings.port}`);
});
